src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessThread.Windows.cs

```diff
@@ -176,5 +176,48 @@ private SafeThreadHandle OpenThreadHandle(int access)
             EnsureState(State.IsLocal);
             return ProcessManager.OpenThread((int)_threadInfo._threadId, access);
         }
+
+        internal bool StateRequiresLazyEvaluation { get; set; }
+
+        /// <summary>
+        /// If the object is populated from ProcessSnapshot,
+        /// ThreadState and ThreadWaitReason need to be lazily evaluated.
+        /// </summary>
+        private ThreadState ThreadStateCore
+        {
+            get
+            {
+                if (StateRequiresLazyEvaluation)
+                {
+                    EnsureStateAndWaitReason();
+                }
+                return _threadInfo._threadState;
+            }
+        }
+
+        private unsafe void EnsureStateAndWaitReason()
+        {
+            // Risk: This method calls NtQueryInformationThread with an *undocumented* THREADINFOCLASS
```

Comment:
It is not ok to use undocumented Windows APIs in this repo (both technical and legal reasons).

Reply:
> It is not ok to use undocumented Windows APIs in this repo (both technical and legal reasons).

no problem